src/mds/Server.cc

```diff
@@ -4357,7 +4351,12 @@ void Server::handle_client_openc(MDRequestRef& mdr)
 
     dout(10) << "adding created_ino and delegated_inos" << dendl;
     ocresp.created_ino = in->inode.ino;
-    ocresp.delegated_inos = mdr->session->delegated_inos;
+
+    // Try to delegate some prealloc_inos to the client, if it needs them
+    // FIXME: Make the ratios tunable
+    if (mdr->session->delegated_inos.size() < (unsigned)g_conf()->mds_client_prealloc_inos / 20)
+      mdr->session->delegate_inos(g_conf()->mds_client_prealloc_inos/10, ocresp.delegated_inos);
+
```

Comment:
The only bit I'm not sure about here is whether this is adequately protected at this point. Do we hold some sort of session-level lock that would protect the delegated/prealloc_inos interval sets in this function? It's not clear to me what locks are held here and I don't think the interval_set code is threadsafe on its own...

@batrick, @ukernel - do you know?

Reply:
SessionMap and Server all through the the global mds lock. It's safe.